### PR TITLE
Implement toSet and toMap in Scala as list methods

### PR DESCRIPTION
**Pull request by odeac**
## Overview
Implement toSet and toMap used for converting lists.

This is needed in PoS implementation of withdraw, where we currently implemented these conversions in Rholang

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3465
https://rchain.atlassian.net/browse/RCHAIN-3466

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: We really should use the ParBuilder to build AST in this test.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: All those tests have the same structure. It feels as if we could reduce them all to:

`checkResult("@0!([1,2,3].toSet())", "@0!(Set(1, 2, 3))")`

Or, equally well, we could implement the test in RhoSpec.

Either way, there's duplication between the tests to be removed.

Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Ok, there are error cases - that's another test util method

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala
Comment (JosephDenman) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1401]: I'm curious what your reasoning is for not implementing `toSet` and `toMap` for all Rholang collections that are scala collections under-the-hood. You should just be able to delegate down to the scala collection converters.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (JosephDenman) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: On one hand, using `ParBuilder` in tests for `Reduce` introduces unnecessary dependencies and violates separation of concerns - testing both the normalizer and the evaluator. On the another, using `ParBuilder` increases readability by reducing preparation. I side with SOC.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala
Comment (JosephDenman) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1415]: This PR would be significantly easier to review if this commit was split into two commits, one adding `toMap` and another adding `toSet`.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (JosephDenman) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: What happens if `toMap` is called on lists like `["a", ("b", 2)]` or `[1, ("b", 2)]`?

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (JosephDenman) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Please add a test that verifies duplicated elements are removed from the source collection when the source collection is transformed into a set.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (JosephDenman) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Please add a test that shows that a list like `[("a", 1), ("a", 2)]` is transformed into a map `{ "a" : 1 }`.

**Review by JosephDenman — CHANGES_REQUESTED** [review]

**Review by KentShikama — APPROVED** [review]
I would like to see tests for toSet/toMap with duplicate entries but otherwise looks good.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Test setup is **not testing** the things used in test setup - it's setting up the stage for the test.

If a test fails in its setup - it's either due to a bug one has just introduced, or due to a change to a cross-cutting concern so drastic that the by-hand test setup would be obsolete and lead to the test no longer testing anything that can happen in production - making the test null in the worst case. It's like drawing a good conclusion from the wrong argument.

Heck, I bet some of the hand-crafted ASTs we have now in tests can't ever happen in production either (they're not sorted).

I don't think we understand separation of concerns the same way. If I need the input par to be normalized *and* sorted - does my calling `normalize` (which sorts underneath) violate SOC? (Hint: it doesn't.)

Hand-crafting complex input data for a test where using prod code in setup would suffice for the test's purpose is terribly misguided, detrimental to test reliability and test maintainability. I can't stress enough how wrong this is.

**Review by ArturGajowy — DISMISSED** [review]
Please see comments above

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: > unnecessary dependencies 

quite the contrary, the fact that this test doesn't use ParBuilder for creating Reducer inputs (which is the only way a reducer input is produced in prod BTW) can lead to false positives in tests (par order can affect execution results / pattern matching; in general: ex falso quodlibet, or GIGO if you will).

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1401]: What would be the expected behaviour for the other collections?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1415]: It already is

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (odeac) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: indeed, there is some refactoring required for the ReduceSpec tests. I'll try to address the duplication.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (odeac) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: I side with Joe here. The ReduceSpec should test only the Reducer and the input of the reducer is the normalised AST, not the Rholang code.

If the unit tests are unreadable and unmaintainable then it's a problem in the API of the unit under test and the tests just point to that issue. Using tools to make the tests readable only hides the actual problem.

Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: > The ReduceSpec **should test only** the Reducer and the input of the reducer is the normalised AST, not the Rholang code.

How can I stress more that **test setup is not testing**? Is that a controversial claim?

> the normalised AST, not the Rholang code.

My claim is, normalization is so complex, that the only **reliable** way to have normalized test input is to use the normalizer. Similarly for parsing. Otherwise we end up doing a lot of test fixing when a detail not relevant to them changes.

Have you read my argument for why pretending that we know how to create the complex test input data is wrong until one cannot write the test otherwise?

---

Consider SRP - each piece of code should only have one reason to change. A test can change for the following reasons:
 1. input data needs to change
 2. the way we call the tested code changes
 3. output data needs to change

When we compare 'by-hand inputs' with 'prod-code inputs', the latter reduces the number of times tests need to change -> we have better, more SRP-y tests.

---

Consider Separation of Concerns - Dijkstra mentions it in the context of `thinking intelligently`, which is to limit the scope of consideration so that each aspect is considered in isolation; at any time we occupy ourselves only with one of the aspects. 

Using 'by-hand inputs' forces test reader & writer to think about how a normalized AST with a peek looks like. Using 'prod-code inputs' allows one to focus on the semantics of peek execution. The rholang source code, **is**, in an ultimate sense, the input to the reducer - and the only form of that input a human being can reason about *intelligently* when we consider results of execution.

So, 'prod-code inputs' leads to better, more SoC-y tests.

---

Granted, there are cases where input data needs to be constructed by hand, like e.g.:
 - making sure the reducer terminates as expected for invalid rholang ASTs (not a requirement, but an example)
 - making sure that the sorter outputs a sorted AST when given an not-sorted one (one could still use a test data generator, although without a *guarantee* that the generated data would be not-sorted)

Testing how a peek reduces though should not depend on how exactly the input AST looks like, as long as it's a normalized AST for a peek. More often than not, hand-writen AST will not be normalized. Further, a change to peek normalization could happen that changes execution results (be it a bug or a purposeful feature) - and the test for peek will not notice until it uses prod normalizer.

Comment (odeac) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: if the test code is difficult to write then we have bad smell. Using tools to hide that smell is not helping at all. Instead we should refactor and eliminate the smell.

If we choose to have Rholang code for tests input then we are  not testing the reducer but instead we are testing the whole interpreter (parser/normalizer/reducer) so that test should not live in ReduceSpec but in some InterpreterSpec.

Now, thinking more about the issue, I would go in the opposite direction and  refactor each method (toSet, toMap etc) by moving them outside the reducer and then unit-test them separately by calling the method apply directly, not by running the reducer.

Then the reducer will only contain one test per method, which would test that the actual method implementation is plugged in correctly.

Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: How is what I'm proposing difficult to write?

`checkResult("@0!([1,2,3].toSet())", "@0!(Set(1, 2, 3))")`

It's even easier than writing all of that in Rholang/RhoSpec.
It's also helping in at least 4 ways: readability, ease of writing, reliability (always uses what's **actually** normalized), and maintainability (less reasons for the test to change / SRP).

Maybe using prod code in setup is not using a tool? For sure it's using the ultimate thing that has a say what the test input should be: the code that produces that input in production.

---

> not testing the reducer but instead we are testing the whole interpreter (parser/normalizer/reducer) so that test should not live in ReduceSpec but in some InterpreterSpec.

I disagree that using prod code in test setup is a test for that prod code too. It does increase the fidelity level of that test though, and prevents testing impossible scenarios - and most tests aren't about those. 

Interpreter could be doing more things in `Interpreter.eval()` than just parsing and normalizing, and then reducing. So using parser + normalizer to test reduction is not testing the interperter. Using interpreter input (however obtained) and calling its method would be.

Moving the tests to InterpreterSpec would be :100: fine for me though, and I like it more than the other solution you proposed, for various reasons (incl. effort needed, test fidelity). Can we do that and close the discussion?

Comment (odeac) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: These tests are doing exactly the same as the other methods tests are doing. So let's keep them like this for now and then we see how to refactor all the ReduceSpec tests in a separate PR.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala
Comment (JosephDenman) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1443]: Calling`toMap` on a map should be a no-op. This implementation throws an error.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala
Comment (JosephDenman) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1435]: Just as `toMap` is only defined on lists when the list elements are tuples, `toMap` should be defined on sets when the set elements are tuples. This is specified in the scala collections library. Please use that as your reference to implement the other collection conversions.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (JosephDenman) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala:1056]: I see you duplicate this function in three separate tests. It should probably be extracted.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (JosephDenman) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Nit: you don't need the extract quotations around `"return report errors in failure cases"`.

**Review by JosephDenman — CHANGES_REQUESTED** [review]
Hi. Second review. Are we in agreement that the `toMap` and `toSet` converters should be defined on a larger set of source collections? I think we should just delegate down to the Scala collection converters, throwing errors whenever Scala would throw an error. What do you think?

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (odeac) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala:1056]: this is not a change of mine. This is just automatic formatting.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1435]: So to sum it up: basically toMap and toSet would be methods of the "Seq" protocol and any collection implementing Seq would implement these. This makes sense to me

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1443]: ok. Do as Scala does makes sense

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: When will this PR come?
What's the outcome of this discussion?
What approach will this PR take?

I'd hate to restart this from scratch, so to me that's a no-no until we reach consensus on the above 3 questions.

To me, implementing the new tests in the/a InterpreterSpec would be an improvement over what's in the PR currently.

Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: As mentioned in the stand up, I'm OK with this PR proceeding as it is now. We should have the above discussion settled though before we write the next test that utilizes rholang AST without a strict necessity.

**Comment by ArturGajowy** [comment]
@odeac feel free to dismiss my review, I've got nothing more to add besides the test inputs approach discussion

**Review by JosephDenman — APPROVED** [review]

**Comment by dckc** [comment]
please, please, PLEASE document all rholang changes (including new methods) in https://github.com/rchain/rchain/tree/dev/rholang/examples

**Comment by ArturGajowy** [comment]
@dckc your repeating the please thrice and capsing it won't make us have more time (or willingness) to do this type of work. A pull request, OTOH, would be helpful.

**Comment by dckc** [comment]
Willingness... Darn. I thought this was agreed and someone just forgot.

**Comment by JosephDenman** [comment]
@dckc You have a point. I missed it in the review.

@odeac Would you mind opening a pull request that adds examples of the `toSet` and `toMap` methods?

**Pull request merged**
